src/cluster/placement/selector/mirrored_custom_groups.go

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package selector
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3/src/cluster/placement"
+	"github.com/m3db/m3/src/x/errors"
+
+	"go.uber.org/zap"
+)
+
+type mirroredCustomGroupSelector struct {
+	instanceIDToGroupID InstanceGroupIDFunc
+	logger              *zap.Logger
+	opts            MirroredCustomGroupOptions
+}
+
+// MirroredCustomGroupOptions configure the selector from NewMirroredCustomGroupSelector
+type MirroredCustomGroupOptions interface {
```

Comment:
Depending on what we decide for `ZoneOptions` - either replace `MirroredCustomGroupOptions` with `ZoneOptions` or blow up `MirroredCustomGroupOptions` with the 4 methods

Reply:
Sg! So after thinking about your comment a bit more, I actually factored this into a common `SelectorOptions` (see discussion below)--lmk what you think of that approach! Still not perfect; I couldn't make non sharded take in just a SelectorOptions for instance.